File: python/src/threads/store/wakes.py
"""The pending_wakes index (spec/schema/store.sql; Gate 1 §2.7.3): one row per running background
child of a branch, inserted in the append of its agent_spawned and deleted in the append of its
agent_finished, or of the parent's parked{kind: child} for it (a parked child is resumed by the
control path). A host resumes a branch with rows, so a child a crash stopped still reports and
wakes its parent. Every append writes them through the index hooks (`wake_rows`);
`threads.reduce.wakes.pending_wakes` is the fold that rebuilds them."""

from collections.abc import Callable, Sequence

from threads.log import BranchId, Event, ParseError, ThreadId
from threads.reduce.wakes import change, pending_wakes
from threads.store.appended import Appended
from threads.store.conn import Conn
from threads.store.sql import text_of

_INSERT = (
    "INSERT INTO pending_wakes (branch_id, child_thread_id) VALUES (?, ?) ON CONFLICT DO NOTHING"
)
_DELETE = "DELETE FROM pending_wakes WHERE branch_id = ? AND child_thread_id = ?"


def wake_rows(conn: Conn, a: Appended) -> ParseError | None:
    """The wake rows one append's events insert and delete: an index hook of every append."""
    for event in a.events:
        found = change(event)
        if found is not None:
            added, child = found
            conn.execute(_INSERT if added else _DELETE, (a.branch_id, child))
    return None


def branches(conn: Conn) -> tuple[tuple[str, ThreadId, BranchId], ...]:
    """(tenant, thread, branch) of every branch with a background child still to report: what
    a host resumes."""
    rows = conn.execute(
        "SELECT DISTINCT b.tenant_id, b.thread_id, w.branch_id FROM pending_wakes w"
        " JOIN branches b ON b.branch_id = w.branch_id"
    ).fetchall()
    return tuple(
        (text_of(tenant), ThreadId(text_of(thread)), BranchId(text_of(branch)))
        for tenant, thread, branch in rows
    )


def rebuild(conn: Conn, read: Callable[[BranchId], Sequence[Event]]) -> None:
    """An index wipe's repair: every row again from the logs alone. `read` is each branch's
    resolved events."""
    conn.execute("DELETE FROM pending_wakes")
    found = conn.execute("SELECT branch_id FROM branches").fetchall()
    for (column,) in found:
        branch = BranchId(text_of(column))
        refold(conn, branch, read(branch))


def refold(conn: Conn, branch: BranchId, events: Sequence[Event]) -> None:
    """One branch's wake rows again from its resolved events: a rebuild's or an import's."""
    conn.execute("DELETE FROM pending_wakes WHERE branch_id = ?", (branch,))
    for child in pending_wakes(events, branch):
        conn.execute(_INSERT, (branch, child))

File: typescript/packages/core/src/store/wakes.ts
import { z } from "zod";
import { BranchId, type KnownEvent, ThreadId } from "../log";
import type { Strict } from "../log/zod-types";
import type { ChainEvent } from "../verify";
import type { SqliteDriver } from "./driver";

// The pending_wakes index (spec/schema/store.sql; Gate 1 §2.7.3): one row per running background
// child of a branch, inserted in the append of its agent_spawned and deleted in the append of its
// agent_finished, or of the parent's parked{kind: child} for it (a parked child is resumed by the
// control path). A host resumes a branch with rows, so a child a crash stopped still reports and
// wakes its parent. The rows follow the replay rule: `pendingWakes` over a branch's own events is
// the fold that rebuilds them.

/** The row an event inserts or deletes, if any. */
function change(
  e: KnownEvent,
): { readonly add: boolean; readonly child: string } | undefined {
  if (e.type === "agent_spawned" && e.data.mode === "background")
    return { add: true, child: e.data.child_thread_id };
  if (e.type === "agent_finished")
    return { add: false, child: e.data.child_thread_id };
  if (e.type === "parked" && e.data.address.kind === "child")
    return { add: false, child: e.data.address.id };
  return undefined;
}

/** The background children of `branch` still waiting to report, in spawn order. */
export function pendingWakes(
  events: readonly KnownEvent[],
  branch: string,
): readonly string[] {
  const rows = new Set<string>();
  for (const e of events) {
    const row = e.branch_id === branch ? change(e) : undefined;
    if (row?.add === true) rows.add(row.child);
    else if (row !== undefined) rows.delete(row.child);
  }
  return [...rows];
}

/** Runs inside an append's transaction, after its events are inserted. */
export function recordWakes(
  db: SqliteDriver,
  branch: string,
  added: readonly ChainEvent[],
): void {
  for (const line of added) {
    const row = line.kind === "event" ? change(line.event) : undefined;
    if (row === undefined) continue;
    db.run(
      row.add
        ? "INSERT OR IGNORE INTO pending_wakes (branch_id, child_thread_id) VALUES (?, ?)"
        : "DELETE FROM pending_wakes WHERE branch_id = ? AND child_thread_id = ?",
      [branch, row.child],
    );
  }
}

const WakeBranch: Strict<{
  tenant_id: z.ZodString;
  thread_id: typeof ThreadId;
  branch_id: typeof BranchId;
}> = z.strictObject({
  tenant_id: z.string(),
  thread_id: ThreadId,
  branch_id: BranchId,
});
export type WakeBranch = z.infer<typeof WakeBranch>;

/** Every branch with a background child still to report: what a host resumes. */
export function wakeBranches(db: SqliteDriver): readonly WakeBranch[] {
  const rows = db.all(
    `SELECT DISTINCT b.tenant_id, b.thread_id, w.branch_id FROM pending_wakes w
      JOIN branches b ON b.branch_id = w.branch_id`,
    [],
  );
  return rows.flatMap((row) => {
    const parsed = WakeBranch.safeParse(row);
    return parsed.success ? [parsed.data] : [];
  });
}

/**
 * An index wipe's repair: every row again from the logs alone. `read` is each branch's
 * resolved events.
 */
export function rebuildWakes(
  db: SqliteDriver,
  read: (branch: string) => readonly KnownEvent[] | undefined,
): void {
  db.run("DELETE FROM pending_wakes", []);
  const branches = db.all("SELECT branch_id FROM branches", []);
  for (const row of branches) {
    const branch = z.object({ branch_id: z.string() }).safeParse(row);
    if (!branch.success) continue;
    const { branch_id } = branch.data;
    for (const child of pendingWakes(read(branch_id) ?? [], branch_id))
      db.run(
        "INSERT INTO pending_wakes (branch_id, child_thread_id) VALUES (?, ?)",
        [branch_id, child],
      );
  }
}
